File: src/main/cc/any_sketch/any_sketch.cc
// Copyright 2020 The Any Sketch Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "src/main/cc/any_sketch/any_sketch.h"

#include <algorithm>
#include <cstdint>
#include <iostream>
#include <type_traits>

#include "absl/base/macros.h"
#include "absl/container/fixed_array.h"
#include "absl/container/flat_hash_map.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "src/main/cc/any_sketch/aggregators.h"
#include "src/main/cc/any_sketch/distributions.h"
#include "src/main/cc/any_sketch/hash_function.h"
#include "src/main/cc/any_sketch/util/macros.h"
#include "src/main/cc/any_sketch/value_function.h"

namespace wfa::any_sketch {

AnySketch::AnySketch(std::vector<std::unique_ptr<Distribution>> indexes,
                     std::vector<ValueFunction> values)
    : indexes_(indexes.size()), values_(values.size()) {
  std::move(indexes.begin(), indexes.end(), indexes_.begin());
  std::move(values.begin(), values.end(), values_.begin());
}

size_t AnySketch::register_size() const { return values_.size(); }

absl::Status AnySketch::AggregateIntoRegister(
    int64_t index, absl::Span<const int64_t> new_values) {
  if (new_values.size() != register_size()) {
    return absl::InvalidArgumentError(
        absl::StrCat("Input has wrong dimension. Expected ", register_size(),
                     " but got ", new_values.size()));
  }
  ABSL_ASSERT(new_values.size() == register_size());

  auto [register_itr, inserted] =
      registers_.try_emplace(index, register_size());
  absl::FixedArray<ValueType>& register_values = register_itr->second;

  if (inserted) {
    std::copy(new_values.begin(), new_values.end(), register_values.begin());
    return absl::OkStatus();
  }

  // Otherwise, merge.
  for (size_t i = 0; i < register_values.size(); ++i) {
    const Aggregator& aggregator = GetAggregator(values_[i].aggregator_type);
    register_values[i] =
        aggregator.Aggregate(register_values[i], new_values[i]);
  }
  return absl::OkStatus();
}

absl::StatusOr<int64_t> AnySketch::GetIndex(
    absl::string_view item, const ItemMetadata& item_metadata) const {
  uint64_t product = 1;
  uint64_t linearized_index = 0;
  for (const std::unique_ptr<Distribution>& distribution : indexes_) {
    ASSIGN_OR_RETURN(int64_t distribution_value,
                     distribution->Apply(item, item_metadata));
    int64_t index_part = distribution_value - distribution->min_value();
    linearized_index = product * linearized_index + index_part;
    product *= distribution->size();
  }
  return linearized_index;
}

absl::Status AnySketch::Insert(absl::Span<const unsigned char> item,
                               const ItemMetadata& item_metadata) {
  absl::string_view item_as_string_view(
      reinterpret_cast<const char*>(item.data()), item.size());

  return Insert(item_as_string_view, item_metadata);
}

absl::Status AnySketch::Insert(uint64_t item,
                               const ItemMetadata& item_metadata) {
  std::array<unsigned char, sizeof(item)> arr{};
  absl::little_endian::Store64(arr.data(), item);
  return Insert(arr, item_metadata);
}

absl::Status AnySketch::Insert(absl::string_view item,
                               const ItemMetadata& item_metadata) {
  ASSIGN_OR_RETURN(int64_t index, GetIndex(item, item_metadata));
  absl::FixedArray<int64_t> new_values(register_size());
  for (size_t i = 0; i < register_size(); ++i) {
    ASSIGN_OR_RETURN(new_values[i],
                     values_[i].distribution->Apply(item, item_metadata));
  }
  return AggregateIntoRegister(index, new_values);
}

absl::Status AnySketch::Merge(const AnySketch& other) {
  // TODO(yunyeng): Check compatibility
  for (const auto& [item, new_values] : other.registers_) {
    RETURN_IF_ERROR(AggregateIntoRegister(item, new_values));
  }
  return absl::OkStatus();
}

absl::Status AnySketch::MergeAll(
    absl::Span<const std::unique_ptr<AnySketch>> others) {
  for (const auto& other : others) {
    RETURN_IF_ERROR(Merge(*other));
  }
  return absl::OkStatus();
}

AnySketch::Iterator& AnySketch::Iterator::operator++() {
  ++pos_;
  return *this;
}

AnySketch::Register AnySketch::Iterator::operator*() const {
  return {pos_->first, pos_->second};
}

bool AnySketch::Iterator::operator!=(const Iterator& other) const {
  return pos_ != other.pos_;
}

AnySketch::Iterator AnySketch::begin() const {
  return Iterator(registers_.begin());
}

AnySketch::Iterator AnySketch::end() const {
  return Iterator(registers_.end());
}

}  // namespace wfa::any_sketch

File: src/main/java/org/wfanet/anysketch/AnySketch.java
// Copyright 2020 The Any Sketch Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.anysketch;

import static com.google.common.collect.ImmutableList.toImmutableList;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import org.wfanet.anysketch.AnySketch.Register;
import org.wfanet.anysketch.distributions.Distribution;

/**
 * A generalized sketch class. This sketch class generalizes the data structure required to capture
 * Bloom filters, HLLs, Cascading Legions, Vector of Counts, and other sketch types. It uses a map
 * of register keys to a register value which is a tuple of numeric values.
 */
public class AnySketch implements Iterable<Register> {
  /** Represents a single register of an AnySketch. */
  public static class Register {

    // Linearized register index
    private long index;

    // Values of the register
    private ImmutableList<Long> values;

    Register(long index, ImmutableList<Long> values) {
      this.index = index;
      this.values = values;
    }

    public long getIndex() {
      return index;
    }

    public ImmutableList<Long> getValues() {
      return values;
    }

    @Override
    public String toString() {
      return "<Register index: " + index + " values: " + values + ">";
    }
  }

  private final ImmutableList<Distribution> indexDistributions;
  private final ImmutableList<ValueFunction> valueFunctions;
  private final Map<Long, List<Long>> registers;

  /**
   * Creates an AnySketch.
   *
   * @param indexDistributions Distributions for determining register index
   * @param valueFunctions Distributions for determining register values
   */
  public AnySketch(List<Distribution> indexDistributions, List<ValueFunction> valueFunctions) {
    this.indexDistributions = ImmutableList.copyOf(indexDistributions);
    this.valueFunctions = ImmutableList.copyOf(valueFunctions);
    this.registers = new HashMap<>();

    long indexesLeft = Long.MAX_VALUE;
    for (Distribution distribution : indexDistributions) {
      indexesLeft /= distribution.getSize();
    }
    Preconditions.checkArgument(indexesLeft > 0, "A register index could possibly exceed 2^63 - 1");
  }

  /** Retries the ValueFunctions that the AnySketch was constructed with. */
  public ImmutableList<ValueFunction> getValueFunctions() {
    return valueFunctions;
  }

  /** Merges a set of values into a register. */
  void aggregateIntoRegister(long index, List<Long> values) {
    Preconditions.checkArgument(values.size() == registerSize());
    registers.merge(index, new ArrayList<>(values), this::aggregateIntoExistingRegister);
  }

  /**
   * Adds `item` to the Sketch.
   *
   * <p>While itemMetadata can contain arbitrary values, certain Distributions may require some keys
   * to be present. In particular, OracleDistributions each require a specific key to exist.
   *
   * <p>It is not an error to include unnecessary keys in itemMetadata--they will simply be ignored.
   *
   * @param item an element to insert into the sketch
   * @param itemMetadata values that Distributions may access
   */
  public void insert(String item, Map<String, Long> itemMetadata) {
    long index = getLinearizedIndex(item, itemMetadata);
    @SuppressWarnings("UnstableApiUsage") // For toImmutableList()
    ImmutableList<Long> values =
        valueFunctions.stream()
            .map(valueFunction -> valueFunction.getDistribution().apply(item, itemMetadata))
            .collect(toImmutableList());
    aggregateIntoRegister(index, values);
  }

  /**
   * Inserts a new item to the AnySketch object with following arguments.
   *
   * @param item Long key
   * @param values Map of values
   */
  public void insert(Long item, Map<String, Long> values) {
    insert(item.toString(), values);
  }

  /**
   * Merges the other sketch into this one. The result is equivalent to sketching the union of the
   * sets that went into this and the other sketch.
   *
   * @param other {@link AnySketch} object
   */
  public void merge(AnySketch other) {
    for (Map.Entry<Long, List<Long>> entry : other.registers.entrySet()) {
      aggregateIntoRegister(entry.getKey(), entry.getValue());
    }
  }

  /**
   * Merges all the other sketches into this one. The result is equivalent to sketching the union of
   * all of the sets.
   *
   * @param others List of {@link AnySketch} objects
   */
  public void mergeAll(Iterable<AnySketch> others) {
    for (AnySketch other : others) {
      merge(other);
    }
  }

  /**
   * Iterates through the registers that the AnySketch object holds.
   *
   * @return Iterator of {@link Register}
   */
  @Override
  public Iterator<Register> iterator() {
    Iterator<Map.Entry<Long, List<Long>>> iterator = registers.entrySet().iterator();
    return new Iterator<Register>() {
      @Override
      public boolean hasNext() {
        return iterator.hasNext();
      }

      @Override
      public Register next() {
        Map.Entry<Long, List<Long>> entry = iterator.next();
        return new Register(entry.getKey(), ImmutableList.copyOf(entry.getValue()));
      }
    };
  }

  private int registerSize() {
    return valueFunctions.size();
  }

  private long getLinearizedIndex(String item, Map<String, Long> itemMetadata) {
    long product = 1L;
    long linearizedIndex = 0;

    for (Distribution distribution : indexDistributions) {
      long indexPart = distribution.apply(item, itemMetadata) - distribution.getMinValue();
      linearizedIndex = product * linearizedIndex + indexPart;
      product *= distribution.getMaxValue() - distribution.getMinValue();
    }
    return linearizedIndex;
  }

  private List<Long> aggregateIntoExistingRegister(List<Long> oldValues, List<Long> newValues) {
    Preconditions.checkState(oldValues.size() == registerSize());
    for (int i = 0; i < registerSize(); i++) {
      long oldValue = oldValues.get(i);
      long newValue = valueFunctions.get(i).getAggregator().aggregate(oldValue, newValues.get(i));
      oldValues.set(i, newValue);
    }
    return oldValues;
  }
}
